### ISSUE TITLE
Protected functions + getters

### ISSUE THREAD

**Issue by peldax** [markrogoyski/math-php#320]
Hello,
Thank you for this amazing library. I am using only small part of it, but its really well done and I really like it.

I have only two small suggestions to maybe improve usability of this lib:

- make the functions protected instead of private:
    - Lets say I am using the Rational class and I don't like the superscript and subscript output. What I would like to do, is to override those two methods to just simply return the numerator and denominator without any conversion, but I cant because of the strictness in access modifier.

- add getters for properties:
    - Again in Rational class, lets say I successfully used your library to make some additions and other mathematical operations with rational numbers. Now I have the result which I want to store in database. I have separate columns for numerator and denominator - but there is currently no way to easily extract those numbers.

What I ended up doing is that I overridden the class only to 
- add getters
- override whole __toString method to actually change only small part of it

Please don't take this as a criticism, I love the lib. Those are only two small suggestions which occurred to me while using it. Maybe you already thought about those and decided not to implement it for some reason. I would like to know your opinion.

Thanks.

peldax

**Comment by markrogoyski** [comment]
Hi @peldax,
Thanks for your interest in MathPHP and your kind words.

Regarding the visibility of the methods, The OOP purist in me wants to argue that the private methods `numeratorToSuperscript` and `denominatorToSubscript` are implementation details and should remain private. They are only called from the public interface method `__toString`. The behavior it sounds you want to override is `__toString`, so creating your own version of that does what you need it to do. It sounds like in your overridden version you could call `parent::__toString` and then do a string replace of the super/sub scripts to regular numbers and be done with it. Even though `numeratorToSuperscript` is private, it is called from a public interface method `__toString`, so you still have access to it in way via `parent::__toString`. And it is better that you don't have direct access to it, because as it is an implementation detail, it could change or even disappear without affecting Rational's public interface, but it would then break your overriding class.

Regarding the getters, I think there are probably some valid reasons someone might want to know what the numerator or denominator is. So it seems like a reasonable request. I'm hesitant to implement this though. My concern is that it requires knowing the private implementation details to really understand the return value. The Rational class actually normalizes the number into three parts: whole number, numerator, and denominator. Someone might construct a Rational number a numerator of 3 and a denominator of 2, and expect `getNumerator` to return 3, but it would actually return 1, since it normalized the fraction's whole part to be 1 and now the fraction is ¹/₂. This could be "solved" with documentation stating the behavior, but considering most people don't read documentation, what would a reasonably smart person expect `getNumerator` to return in this case after having constructed it with a 3?

This is easy to do, so if you think this is an important feature that adds value, I'm OK with it. I think it may create confusion though, so if you think the benefits outweigh the potential confusion it brings, then I can add it. Let me know what you think.

Thanks,
Mark

**Comment by Beakerboy** [comment]
I agree with Mark on the superscript/subscript details, but I think getters are a good idea. If someone wants to accurately “cast” a RationalNumber into something else, they need all the data. Converting to a float and then into the “something else” will inherently lose the precision that this class was built for.

**Comment by peldax** [comment]
About the access modifiers - you are right. Although I don't agree with the superscript/subscript format, it is how it is, and private converting methods should remain private as they are implementation details. I was too concerned about desired behaviour, that I missed the design decisions.

About the getters - I understand the confusion you are talking about. I was thinking about it and came to same suggestion as @Beakerboy in comment above - Create one single method which returns array.

This way you have two way conversion from object to array using array unpacking.
```
$number = new Rational(1, 2, 3);
$data = $number->getData();
$newNumber = new Rational(..$data);
```

**Comment by Beakerboy** [comment]
Sounds like you know what you’re doing. Go ahead and type it up and submit a pull request. We (the 3 of us) can continue to discuss and make small changes if needed. I’d just add the unit testing assertion to the toString test since both output will be equally normalized.

**Comment by markrogoyski** [comment]
I suggest adding `getData` and individual element getters `getNumerator`, `getDenominator`, and `getWholeNumber` (think of a better name for that one that doesn't confuse the idea of getting all of the numbers).

**Comment by Beakerboy** [comment]
Maybe...getWholePart(), getNumeratorPart(), getDenominatorPart()

**Comment by peldax** [comment]
Should I make the PR only for Rational class, or any other class as well? 

I am currently using only Rational class, so it sufficient for me, but inconsistent with other classes.

**Comment by Beakerboy** [comment]
We’re more than happy to have as much help from anybody on anything.

**Comment by markrogoyski** [comment]
How about as a first pull request, just focus on the changes your interested in for the Rational number class. Take a look at the Wiki documentation in this project and some other classes to see how we write unit tests. If you have any questions, don't hesitate to ask. Thanks for offering to contribute!

**Comment by markrogoyski** [comment]
I've added the basic getters to the Rational class in the develop branch. It will go out with the next release.

**Comment by peldax** [comment]
Thanks. Although its been half a year, I found really no time to create proper PR.

**Comment by markrogoyski** [comment]
This has been released in version 0.57.0: https://github.com/markrogoyski/math-php/releases/tag/v0.57.0
